### PR TITLE
Add _WD_GetFreePort

### PR DESCRIPTION
**Pull request by Danp2**
### Proposed changes

Add function _WD_GetFreePort

### Checklist

Put an `x` in the boxes that apply. If you're unsure about any of them, don't hesitate to ask. We are here to help!<br>
This is simply a reminder of what we are going to look for before merging your code.

- [x] I have read and noticed the [CODE OF CONDUCT](https://github.com/Danp2/au3WebDriver/blob/master/docs/CODE_OF_CONDUCT.md) document
- [x] I have read and noticed the [CONTRIBUTING](https://github.com/Danp2/au3WebDriver/blob/master/docs/CONTRIBUTING.md) document
- [ ] I have added necessary documentation or screenshots (if appropriate)

### Types of changes

Please check `x` the type of change your PR introduces:

- [ ] Bugfix (change which fixes an issue)
- [x] Feature (change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (functional, structural)
- [ ] Documentation content changes
- [ ] Other (please describe)

### What is the current behavior?

No easy means to identify an alternate port number that is available for use

### What is the new behavior?

_WD_GetFreePort can be called to obtain a free port number within a defined range

### System under test

Please complete the following information.

- OS: [e.g. Windows 10]
- OS Arch.: [e.g. X64]
- Browser [e.g. firefox]
- Browser version [e.g. 96.0.3]

### CODE REVIEW

File: wd_helper.au3 [hunk 1]
```diff
@@ -2906,6 +2906,71 @@ Func _WD_JsonActionKey($sType, $sKey, $iSuffix = Default)
...
+	Local $aPorts = __WinAPI_GetTcpTable()
+
+	If Not @error Then
+		For $iPort = $iMinPort To $iMaxPort
+			_ArraySearch($aPorts, $iPort, Default, Default, Default, Default, Default, 3)
+			If @error = 6 Then Return $iPort
+		Next
+	Endif
+
+	Return 0
```
Comment (mlipok) [inline wd_helper.au3:2935]: What about reporting `@error` ?

**Comment by mlipok** [comment]
I also start thinking about:
Should we check if port providen by user is free or used and return specific error instead runing webdriver exe to check what will happen ?

File: wd_helper.au3 [hunk 2]
```diff
@@ -2906,6 +2906,71 @@ Func _WD_JsonActionKey($sType, $sKey, $iSuffix = Default)
...
+EndFunc   ;==>_WD_GetFreePort
+
+Func __WinAPI_GetTcpTable()
+    ;funkey 2012.12.14
+	;https://www.autoitscript.com/forum/topic/146671-getextendedtcptable-get-netstat-information/?tab=comments#comment-1038649
```
[the remark this answers is not in the document]
Comment (Danp2) [reply wd_helper.au3:2940]: It's a tab vs spaces issue. I'll fix it.

File: wd_helper.au3 (hunk 1 above)
Comment (Danp2) [reply wd_helper.au3:2935]: I'll update to use __WD_Error like other functions.

**Comment by Danp2** [comment]
> Should we check if port providen by user is free or used and return specific error instead runing webdriver exe to check what will happen ?

IDK. I guess that could be a nice option to add. I would think you could do something like this --
```autoit
If Not _WD_GetFreePort(9515) Then
    ; Take action here due to unavailable port
EndIf
```

However, this is problematic because you would likely want to add this check in `_WD_Startup` and `_WD_GetFreePort` doesn't exist in wd_core.au3. I specifically placed it in wd_helper.au3 because it isn't "core" webdriver functionality IMO.

File: wd_helper.au3 [hunk 3]
```diff
@@ -2922,9 +2922,12 @@ EndFunc   ;==>_WD_JsonActionKey
 ; Link ..........:
 ; Example .......: No
 ; ===============================================================================================================================
-Func _WD_GetFreePort($iMinPort, $iMaxPort)
+Func _WD_GetFreePort($iMinPort = Default, $iMaxPort = Default)
 	Local Const $sFuncName = "_WD_GetFreePort"
 	Local $iResult = 0, $iErr = $_WD_ERROR_NotFound
+
+	If $iMaxPort = Default Then $iMaxPort = ($iMinPort = Default) ? 65000 : $iMinPort
+	If $iMinPort = Default Then $iMinPort = 64000
```
Comment (mlipok) [inline wd_helper.au3:2930]: I propose to defult values
```autoit
If $iMinPort = Default Then $iMinPort = $_WD_PORT
If $iMaxPort = Default Then $iMaxPort = ($iMaxPort = Default) ? ($iMinPort + 100) : $iMaxPort
```

Comment (Danp2) [reply wd_helper.au3:2930]: @mlipok Appreciate the feedback. However, I have some concerns regarding your proposed changes --

- This function will be used in other areas (such as BiDi), so it is intended to be generic in nature and not specifically related to the port used for web driver communication

- If $_WD_Port <> 0, then that means the user has already assigned a value using _WD_Option. It makes more sense to me if these were the order of events --
```
1. Obtain a port number using _WD_GetFreePort (optional)
2. Set the port number using _WD_Option
3. Call _WD_Startup
```
- I don't see how your 2nd line modifying $iMaxPort can even function properly. Care to explain?

Comment (mlipok) [reply wd_helper.au3:2930]: Ok.
Should we also add som usage to _wd_demo.au3 ?

**Comment by mlipok** [comment]
Should we add some example usage to wd_demo.au3 ?

**Comment by Danp2** [comment]
I don't think so. We can't cover every possible option in wd_demo. I think this could be a good FAQ for the wiki.

**Pull request merged**